docs/style-spec/expressions.md

```diff
@@ -0,0 +1,195 @@
+
+A zoom function must be of the following form (see [Curves](#Curves) below):
+
+```js
+{
+  "expression": [ "curve", interpolation, ["zoom"], ... ]
+}
+```
+
+Or:
+
+```js
+{
+  "expression": [ "coalesce", [ "curve", interpolation, ["zoom"], ... ], ... ]
+}
+```
+
+(`["zoom"]` may not appear anywhere else in the expression.)
+
+
+### Example: a zoom-only function.
+
+```js
+{
+  "circle-radius": {
+    "expression": [
+      "curve", "linear", ["zoom"],
+      // zoom is 5 (or less) -> circle radius will be 1px
+      // zoom is 10 (or greater) -> circle radius will be 2px
+      5, 1, 10, 2
+    ]
+  }
+}
+```
+
+### Example: a zoom-and-property function
+
+Using property functions as the output value for one or more zoom stops allows
+the appearance of a map feature to change with both the zoom level _and_ the
+feature's properties.
+
+```js
+{
+  "circle-radius": {
+    "expression": [
+      "curve", "linear", ["zoom"],
+
+      // zoom is 0 and "rating" is 0 -> circle radius will be 0px
+      // zoom is 0 and "rating" is 5 -> circle radius will be 5px
+      0, ["number", [ "get", "rating" ],
+      // zoom is 20 and "rating" is 0 -> circle radius will be 4 * 0 = 0px
+      // zoom is 20 and "rating" is 5 -> circle radius will be 4 * 5 = 20px
+      10, [ "*", 4, [ "number", [ "get", ["properties"] ] ] ]
+    ]
+  }
+}
+```
+
+## Property Expressions
+
+Property expressions are represented using a Lisp-like structured syntax tree.
+
+### Types
+
+Every expression evaluates to a value of one of the following types.
+
+- `Null`
+  - Literal: `null`
+- `String`
+  - Literal: `"my string value"`
+- `Number`
+  - Literal: `1729`
+- `Boolean`
+  - Literal: `true` or `false`
+- `Color`
+- `Object`
+  - Literal: `["literal", { "key": value, "key2": value2 }]`
+- `Array<T, N>`: an array of N values of type T
+  - Literal: `["literal", [v0, v1, ...]]`
+- `Array<T>`: a dynamically sized array of values of type `T`
+- `Array`: equivalent to `Array<Value>`
+- `Value`: A "variant" type representing the set of possible values retrievable from a feature's `properties` object (`Null | String | Number | Boolean | Object | Array`)
+- `Error`: a subtype of all other types. Used wherever an expression is unable to return the appropriate supertype. Carries diagnostic information such as an explanatory message and the expression location where the Error value was generated.
+
+### Constants:
+- `[ "ln2" ] -> Number`
+- `[ "pi" ] -> Number`
+- `[ "e" ] -> Number`
+
+### Variable binding:
+- `[ "let", name1: String, e1, name2: String, e2, ..., e_result ]` - Bind expression `e1` to the string `name1`, `e2` to `name2`, etc., before evaluating `e_result`.  The bound expressions may be referenced within `e_result` with `[ name1 ]`, `[ name2 ]`, etc. (E.g.: `["let", "a", 1, "b", ["number", ["get", "blah"]], [ "+", ["a"], ["b"] ]`.)
```

Comment:
🤦‍♂️ yeah.  `["ref", name1]`? `["var", name1]`?  Less concise, but actually might allow us to avoid having special logic for var reference expressions.